File: AegisSpark/abe-eac/native/src/path_util.cpp
#include "abe_spark/path_util.hpp"

#include <algorithm>
#include <cctype>

namespace abe_spark {
namespace {

std::string stripTrailingSlash(const std::string& s)
{
	if (s.size() > 1 && s.back() == '/') {
		return s.substr(0, s.size() - 1);
	}
	return s;
}

} // namespace

std::string PathUtil::normalize(const std::string& hdfsUri)
{
	std::string out = hdfsUri;
	while (!out.empty() && std::isspace(static_cast<unsigned char>(out.front()))) {
		out.erase(out.begin());
	}
	while (!out.empty() && std::isspace(static_cast<unsigned char>(out.back()))) {
		out.pop_back();
	}
	if (out.empty()) return out;
	return stripTrailingSlash(out);
}

bool PathUtil::isUnderRoot(const std::string& target, const std::string& root)
{
	const std::string t = normalize(target);
	const std::string r = normalize(root);
	if (r.empty() || t.empty()) return false;
	if (t == r) return true;
	if (t.size() <= r.size()) return false;
	if (t.compare(0, r.size(), r) != 0) return false;
	return t[r.size()] == '/';
}

bool PathUtil::isUnderAnyRoot(const std::string& target, const std::vector<std::string>& roots)
{
	for (size_t i = 0; i < roots.size(); ++i) {
		if (isUnderRoot(target, roots[i])) return true;
	}
	return false;
}

bool PathUtil::ticketBindsTarget(const std::string& ticketPath, const std::string& writeTarget)
{
	const std::string t = normalize(ticketPath);
	const std::string w = normalize(writeTarget);
	if (t.empty() || w.empty()) return false;
	return t == w || isUnderRoot(w, t);
}

} // namespace abe_spark

File: sgx-pyspark/sgx_pyspark/path_util.py
"""HDFS URI 路径规范化与授权前缀匹配（与 ABE-Spark PathUtil 语义对齐）。"""

from __future__ import annotations


def normalize(hdfs_uri: str) -> str:
    uri = hdfs_uri.strip()
    if not uri.startswith("hdfs://"):
        return uri.rstrip("/")
    scheme_end = uri.find("/", 7)
    if scheme_end < 0:
        return uri.rstrip("/")
    path = uri[scheme_end:]
    while "//" in path:
        path = path.replace("//", "/")
    return uri[:scheme_end] + path.rstrip("/")


def is_under_root(target: str, root: str) -> bool:
    t = normalize(target)
    r = normalize(root)
    if not r:
        return False
    if t == r:
        return True
    prefix = r if r.endswith("/") else r + "/"
    return t.startswith(prefix)


def is_under_any_root(target: str, roots: list[str]) -> bool:
    return any(is_under_root(target, root) for root in roots)


def ticket_binds_target(ticket_path: str, write_target: str) -> bool:
    t = normalize(ticket_path)
    w = normalize(write_target)
    if not t or not w:
        return False
    return t == w or is_under_root(w, t)
